Common/Orders/Fills/FillModel.cs

```diff
@@ -106,13 +113,202 @@ public virtual Fill Fill(FillModelParameters parameters)
+                // we haven't received fill requests for all the orders in the group
+
+                if (!pendingFillsParameters.Any(x => x.Order.Id == order.Id))
+                {
+                    pendingFillsParameters.Add(Parameters);
+
+                    if (pendingFillsParameters.Count < groupOrderManager.Count)
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        private ComboFill FillComboOrder(FillModelParameters parameters)
+        {
+            var order = parameters.Order;
+            if (order.GroupOrderManager == null)
+            {
+                // This should never happen
+                throw new ArgumentException("FillModel.FillComboOrder(): The order is not a grouped order");
+            }
+
+            if (!CheckIfComboOrderIsReadyToFill(order))
+            {
+                return ComboFill.Empty;
+            }
+
+            List<OrderEvent> orderEvents;
+
+            switch (order.Type)
+            {
+                case OrderType.ComboMarket:
+                    orderEvents = ComboMarketFill(order as ComboMarketOrder);
+                    break;
+                case OrderType.ComboLimit:
+                    orderEvents = ComboLimitFill(order as ComboLimitOrder);
+                    break;
+                case OrderType.ComboLegLimit:
+                    orderEvents = ComboLegLimitFill(order as ComboLegLimitOrder);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException();
+            }
+
+            if (orderEvents.Count > 0)
+            {
+                // orders are filled, we can clear the pending orders table
+                _pendingGroupedOrdersByManagerId.Remove(order.GroupOrderManager.Id);
+            }
+
+            return new ComboFill(orderEvents);
+        }
+
+        /// <summary>
+        /// Default combo market fill model for the base security class. Fills at the last traded price for each leg.
+        /// </summary>
+        /// <param name="order">Order to fill</param>
+        /// <returns>Order fill information detailing the average price and quantity filled for each leg. If any of the fills fails, none of the orders will be filled and the returned list will be empty</returns>
+        public virtual List<OrderEvent> ComboMarketFill(ComboMarketOrder order)
+        {
+            // if this was called, we are ready to fill the combo order and event the current one should be in the pending orders table
+            var fills = _pendingGroupedOrdersByManagerId[order.GroupOrderManager.Id]
+                .Select(x => InternalMarketFill(x.Security, x.Order, x.Order.Quantity * order.GroupOrderManager.Quantity));
+
+            if (fills.Any(x => x.Status != OrderStatus.Filled))
+            {
+                return new List<OrderEvent>();
+            }
+
+            return fills.ToList() ;
```

Comment:
- Whops! `Any` & `ToList` looping twice over the Select above, ideally we do this once. 
- Could even be more lazy, using a normal for loop the first order even we get that isn't filled causes us to break/stop.
- Should avoid creating a new list we don't need `new List<OrderEvent>()`

Same comments for the others bellow 👍